pubsub/azurepubsub/azurepubsub.go

```diff
@@ -0,0 +1,438 @@
+// Copyright 2018 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package azurepubsub provides an implementation of pubsub that uses Azure Service Bus Topics and Subscriptions
+// PubSub.
+//
+// It exposes the following types for As:
+// Topic: *servicebus.Topic
+// Subscription: *servicebus.Subscription
+package azurepubsub // import "gocloud.dev/pubsub/azurepubsub"
+
+import (	
+	"gocloud.dev/gcerrors"
+	"context"
+	"fmt"	
+	"pack.ag/amqp"
+	"runtime"
+	"sync"
+	"time"
+	
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+	"gocloud.dev/internal/useragent"
+
+	"github.com/Azure/azure-amqp-common-go/cbs"
+	"github.com/Azure/azure-amqp-common-go/rpc"
+	"github.com/Azure/azure-amqp-common-go/uuid"
+	"github.com/Azure/azure-service-bus-go"
+)
+
+const (
+	completedStatus  = "completed"	
+	listenerTimeout = 1 * time.Second
+	rpcTries = 5
+	rpcRetryDelay = 1 * time.Second
+)
+
+type topic struct {		
+	name string
+	ns *servicebus.Namespace	
+}
+
+// TopicOptions provides configuration options for an Azure SB Topic.
+type TopicOptions struct {}
+
+// OpenTopic opens the topic on Azure ServiceBus PubSub for the given topicName.
+func OpenTopic(ctx context.Context, topicName string, connString string, opts *TopicOptions) (*pubsub.Topic) {
```

Comment:
No longer needed. OpenTopic changed to accept a SB Topic which holds the AMQP connection for sending. This removes the need for me to manage senders since I am now using topic.Send

func OpenTopic(ctx context.Context, sbTopic *servicebus.Topic, opts *TopicOptions) (*pubsub.Topic) {
	t := openTopic(ctx, sbTopic)	
	return pubsub.NewTopic(t)
}`
`